kornia/geometry/plane.py

```diff
@@ -33,74 +34,96 @@ def __str__(self) -> str:
+            # TODO: improve tests
+            KORNIA_CHECK_SHAPE(p0, ["*", "2"])
+            KORNIA_CHECK(p0.shape == p1.shape)
             # TODO: implement `.unitOrthonormal`
-            normal = normalize((p1 - p0), p=2, dim=-1)
-            offset = -batched_dot_product(p0, normal)
-            return Hyperplane(normal, offset)
+            normal2d = normalized(p1 - p0)
+            offset2d = -batched_dot_product(p0, normal2d)
+            return Hyperplane(wrap(normal2d, Vector3), wrap(offset2d, Scalar))
         # 3d case
-        # KORNIA_CHECK_SHAPE(p0, ["B", "3"])  # FIXME: resolve shape bugs. @edgarriba
-        # KORNIA_CHECK_SHAPE(p1, ["B", "3"])  # FIXME: resolve shape bugs. @edgarriba
-        # KORNIA_CHECK_SHAPE(p2, ["B", "3"])  # FIXME: resolve shape bugs. @edgarriba
-        v0, v1 = (p2 - p1), (p1 - p0)
+        KORNIA_CHECK_SHAPE(p0, ["*", "3"])
+        KORNIA_CHECK(p0.shape == p1.shape)
+        KORNIA_CHECK(p1.shape == p2.shape)
+        v0, v1 = (p2 - p0), (p1 - p0)
         normal = v0.cross(v1)
         norm = normal.norm(-1)
-        # TODO: use where
+
         # https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Geometry/Hyperplane.h#L108
-        # general case
-        normal = normal / (norm + 1e-6)
+        def compute_normal_svd(v0, v1):
+            # NOTE: for reason TensorWrapper does not stack well
+            m = stack((unwrap(v0), unwrap(v1)), -2)  # Bx2x3
+            _, _, V = _torch_svd_cast(m)  # kornia solution lies in the last row
+            return wrap(V[..., -1, :], Vector3)  # Bx3
+
+        normal_mask = norm <= v0.norm(-1) * v1.norm(-1) * 1e-6
+        normal = where(normal_mask, compute_normal_svd(v0, v1), normal / (norm + 1e-6))
         offset = -batched_dot_product(p0, normal)
-        return Hyperplane(normal, offset)
+
+        return Hyperplane(wrap(normal, Vector3), wrap(offset, Scalar))
 
 
 # TODO: factor to avoid duplicated from line.py
 # https://github.com/strasdat/Sophus/blob/23.04-beta/cpp/sophus/geometry/fit_plane.h
-def fit_plane(points: Tensor) -> Hyperplane:
+def fit_plane(points: Vector3) -> Hyperplane:
     """Fit a plane from a set of points using SVD.
 
     Args:
         points: tensor containing a batch of sets of n-dimensional points. The  expected
-            shape of the tensor is :math:`(B, N, D)`.
+            shape of the tensor is :math:`(N, D)`.
 
     Return:
         The computed hyperplane object.
     """
-    KORNIA_CHECK_SHAPE(points, ["B", "N", "D"])
+    KORNIA_CHECK_SHAPE(points, ["N", "D"])
 
     mean = points.mean(-2, True)
     points_centered = points - mean
 
     # NOTE: not optimal for 2d points, but for now works for other dimensions
-    _, _, V = _torch_svd_cast(points_centered)
-    V = V.transpose(-2, -1)
+    U, S, V = _torch_svd_cast(points_centered)
 
     # the first left eigenvector is the direction on the fited line
-    direction = V[..., 0, :]  # BxD
+    direction = V[..., :, -1]  # BxD
```

Comment:
Review `_torch_svd_cast` result. Internally we transpose, which should be avoided